[issue #1201] MQTT#Disconnected event and MQTT stability improvements

TD-er opened the pull request:
While looking for issues regarding DeepSleep + MQTT and issues #1194 & #1195, I came across a number of other issues related to high CPU load and WiFi disconnects when a MQTT controller cannot reach the MQTT broker or is set to be no longer enabled.

These fixes improve CPU load and WiFi stability and also send the following events:
- `MQTT#Connected` 
- `MQTT#Disconnected`
- `MQTTimport#Connected` 
- `MQTTimport#Disconnected`

The MQTTimport plugin still has some major issues which have to be taken care of, so #1195 is not fixed yet.

TD-er committed:
[issue #1201] MQTT#Disconnected event and MQTT stability improvements

While looking for issues regarding DeepSleep + MQTT and issue #1195, I came across a number of other issues related to high CPU load and WiFi disconnects when a MQTT controller cannot reach the MQTT broker or is set to be no longer enabled.

These fixes improve CPU load and WiFi stability and also send the following events:
- `MQTT#Connected` 
- `MQTT#Disconnected`
- `MQTTimport#Connected` 
- `MQTTimport#Disconnected`

The MQTTimport plugin still has some major issues which have to be taken care of, so #1195 is not fixed yet.

TD-er committed:
[issue #1201] MQTT#Disconnected event and MQTT stability improvements

While looking for issues regarding DeepSleep + MQTT and issue #1195, I came across a number of other issues related to high CPU load and WiFi disconnects when a MQTT controller cannot reach the MQTT broker or is set to be no longer enabled.

These fixes improve CPU load and WiFi stability and also send the following events:
- `MQTT#Connected` 
- `MQTT#Disconnected`
- `MQTTimport#Connected` 
- `MQTTimport#Disconnected`

The MQTTimport plugin still has some major issues which have to be taken care of, so #1195 is not fixed yet.

TD-er commented:
If someone can test this one first, since it is too far into the night to merge this and get away with it ;)

waspie commented:
damn, and i just recompiled and updated my units for the night :\
thanks for doing this and I wonder if what you fixed is the root of my issues causing the disconnects in the first place...
I doubt i'll be able to mess with this until tomorrow

as far as high load, it wasn't too long ago that if i botched the ip of the mqtt broker the esp would be useless until i did a reset on it (not reboot) because it would be continually trying to connect to mqtt.

TD-er commented:
Yeah, took me a while.
Thing is that it's now past 3am and I guess it's better to merge it tomorrow after a cup of coffee.
I would say a good night's sleep, but that's not gonna happen I'm afraid... Still thinking about some issues I saw when browsing the code.

waspie commented:
I appreciate the fixes and the mqtt events you've created. they've been real helpful to me

waspie commented:
i just edited an mqtt import on an esp running code from a few days ago and it went non responsive so I'd like to test this but I don't know how to download this change. sorry, i'm not well versed on github, i know i know.

psy0rz committed:
automaticly updated release notes for mega-20180330

v-a-d-e-r commented:
Sending data after deepsleep still not working. :-(

TD-er commented:
@v-a-d-e-r Any idea what might go wrong?
Maybe send me a log, or the settings?
If the DHCP does not work properly, it goes back to sleep again.

v-a-d-e-r commented:
That could be the problem...!?

From the log:
.
.
.
wifi : connecting ....bla bla
.
From here all log entries below are missing now!
(They were there before Mar 25.)
.
event: time#initialized
event: wifi#changedaccesspoint
event: wifi#connected
wifi : connected! ip: ....
event: time#set
.
.
mqtt: connected to broker ....
mqtt: subscribed to ....

With disabled deep sleep the log entries appear! Hope it helps a bit...

UPDATE:
Wow! That is new. The worst we can see in the log:
Exception (0)....
panic
WDT reset

WTF is going on there??? Never saw that...

martinusGH committed:
ESP32: Fix Wifi Disconnect reason

TD-er committed:
[Build ESP32] Fix for build issues between ESP32 and ESP8266

There is a problem with the PlatformIO builder since the last update of @mvdbro
It's an issue with the parsing of `#ifdef ESP32`.

It is very well explained here: [platformio/platformio-core#801 (comment)](https://github.com/platformio/platformio-core/issues/801#issuecomment-252447178)

In short, the event handling functions for the wifi have been moved to a separate .h file, since those files will be parsed properly by the Library Dependency Finder (LDF) in PlatformIO 3.x

TD-er committed:
[ESP32] Build errors/warnings

TD-er committed:
Merge pull request #1212 from TD-er/bugfix/build_issues_ESP32_ESP8266

[Build ESP32] Fix for build issues between ESP32 and ESP8266

TD-er committed:
Merge remote-tracking branch 'origin/feature/mqtt_disconnect_event' into feature/mqtt_disconnect_event

TD-er committed:
[deepsleep] Allow for some minimum wifi connect time

In Deepsleep mode the wifi connect time is now minimum of 6 seconds.
Also the first loop after connect, the 'awake' timer will reset and start over. Without WiFi connection the set 'awake' timer starts at the moment the node boots.

Grovkillen commented:
https://www.letscontrolit.com/wiki/index.php/Tutorial_Rules#.28System.29_events

Documented and closed .
